calico_rkt/calico_rkt.py

```diff
@@ -14,387 +14,485 @@
+    
+    def delete(self):
+        """Handles CNI_CMD_DELETE requests.
+
+        Remove this pod from Calico networking.
+
+        :return: None.
+        """
+        _log.info('Deleting pod %s' % self.container_id)
+
+        # Step 1: Remove any IP assignments.
+        self._release_ip()
+
+        # Step 2: Get the Calico endpoint for this workload. If it does not
+        # exist, log a warning and exit successfully.
+        endpoint = self._get_endpoint()
+        if not endpoint:
+            _log.info("No more to do since endpoint does not exist.",
+                       self.container_id)
+            sys.exit(0)
+
+        # Step 3: Delete the veth interface for this endpoint.
+        _log.info("Removing veth for endpoint %s", endpoint.name)
+        netns.remove_veth(endpoint.name)
+
+        # Step 4: Delete the Calico endpoint.
+        self._remove_endpoint()
+
+        # Step 5: Delete any profiles for this endpoint
+        # TODO: This is racey and needs further thinking.
+        self._remove_profile(self.network_name)
+
+        _log.info('Finished deleting pod %s' % self.container_id)
+
+    def _assign_ip(self):
+        """Assigns and returns an IPv4 address using the IPAM plugin
+        specified in the network config file.
+
+        :return: IPAddress - The assigned IP address.
+        """
+        # TODO - Handle errors thrown by IPAM plugin
+        self.ipam_result = self._call_ipam_plugin()
+        _log.debug("IPAM plugin result: %s", self.ipam_result)
 
+        try:
+            # Load the response and get the assigned IP address.
+            self.ipam_result = json.loads(self.ipam_result)
+        except ValueError:
+            _log.exception("Failed to parse IPAM response, exiting")
+            sys.exit(1)
 
-def delete(args):
-    """
-    Cleanup after a pod.
+        try:
+            assigned_ip = IPAddress(self.ipam_result["ip4"]["ip"])
+        except KeyError:
+            _log.error("IPAM plugin did not return an IPv4 address.")
+            sys.exit(1)
+        except (AddrFormatError, ValueError):
+            _log.error("Invalid IP address %s", self.ipam_result["ip4"]["ip"])
+            sys.exit(1)
 
-    :param args: dict of values to pass to other functions (see: validate_args)
-    """
-    container_id = args['container_id']
-    net_name = args['name']
+        return assigned_ip
 
-    _log.info('Deleting pod %s' % container_id)
+    def _release_ip(self):
+        """Releases the IP address(es) for this container using the IPAM plugin
+        specified in the network config file.
 
-    # Remove the profile for the workload.
-    _container_remove(hostname=HOSTNAME,
-                      orchestrator_id=ORCHESTRATOR_ID,
-                      container_id=container_id)
+        :return: None.
+        """
+        # TODO - Handle errors thrown by IPAM plugin
+        _log.debug("Releasing IP address with IPAM plugin")
+        try:
+            result = self._call_ipam_plugin
```

Comment:
You're not actually calling the the ipam plugin here - need some `( )` up in here.